xarray/core/computation.py

```diff
@@ -0,0 +1,675 @@
+import collections
+import functools
+import itertools
+import operator
+import re
+
+from . import ops
+from .alignment import deep_align
+from .merge import merge_coords_without_align
+from .utils import is_dict_like
+from .pycompat import dask_array_type, OrderedDict, basestring
+
+
+_DEFAULT_FROZEN_SET = frozenset()
+
+# see http://docs.scipy.org/doc/numpy/reference/c-api.generalized-ufuncs.html
+DIMENSION_NAME = r'\w+'
+CORE_DIMENSION_LIST = '(?:' + DIMENSION_NAME + '(?:,' + DIMENSION_NAME + ')*)?'
+ARGUMENT = r'\(' + CORE_DIMENSION_LIST + r'\)'
+ARGUMENT_LIST = ARGUMENT + '(?:,' + ARGUMENT + ')*'
+SIGNATURE = '^' + ARGUMENT_LIST + '->' + ARGUMENT_LIST + '$'
+
+
+def safe_tuple(x):
+    # type: Iterable -> tuple
+    if isinstance(x, basestring):
+        raise ValueError('cannot safely convert %r to a tuple')
+    return tuple(x)
+
+
+class Signature(object):
```

Comment:
This is already a [class](https://docs.python.org/3/library/inspect.html#inspect.Signature) in Python3 FYI